### PR TITLE
Allow in-viewport width-expansion when reflow is not possible

### PR DESCRIPTION
**Pull request by glevitzky**
This change will allow elements within the viewport to expand their widths via attemptChangeSize, if the following conditions hold:

1. The element has no siblings in the DOM
2. The parent of the element is at least as wide as the new width.

Tests will be included once basic implementation is approved.

### CODE REVIEW

File: src/service/resources-impl.js [hunk 1]
```diff
@@ -1585,6 +1593,26 @@ export class Resources {
...
   }
 
+  /**
+   * Returns true if reflow will not be caused by expanding the given element's
+   * width to the given amount.
+   * @param {!Element} element
+   * @param {number} width
+   * @return {boolean}
+   */
+  widthOnlyReflowFreeExpansion_(element, width) {
```
Comment (lannka) [inline src/service/resources-impl.js]: naming nit: isWidthOnly...

Comment (glevitzky) [reply src/service/resources-impl.js]: Done.

File: src/service/resources-impl.js [hunk 2]
```diff
@@ -1585,6 +1593,26 @@ export class Resources {
...
+   * @return {boolean}
+   */
+  isWidthOnlyReflowFreeExpansion_(element, width) {
+    const parent = element && element.parentElement;
+    // If the element has siblings, it's possible that a width-expansion will
+    // cause some of them to be pushed down.
+    if (!parent || parent.childElementCount > 1) {
+      return false;
+    }
+    const parentWidth = parseInt(computedStyle(this.win, parent).width, 10);
```
Comment (jridgewell) [inline src/service/resources-impl.js]: This causes a measure in the middle of a mutation phase.

Comment (glevitzky) [reply src/service/resources-impl.js]: @jridgewell Is there a way of getting the width of this element without causing a measure? If not, I guess I can only return true here if the parent has a fixed layout (in which case I can just grab the width attribute).

Comment (jridgewell) [reply src/service/resources-impl.js]: If it's an amp-element, calling `implementation.getLayoutWidth` would work. For regular elements, there's no way to do it.

Comment (glevitzky) [reply src/service/resources-impl.js]: Removed call to getComputedStyle.

File: src/service/resources-impl.js [hunk 3]
```diff
@@ -1585,6 +1595,27 @@ export class Resources {
...
 
+  /**
+   * Returns true if reflow will not be caused by expanding the given element's
+   * width to the given amount.
+   * @param {!Element} element
+   * @param {number} width
+   * @return {boolean}
+   */
+  isWidthOnlyReflowFreeExpansion_(element, width) {
+    const parent = element && element.parentElement;
```
Comment (jridgewell) [inline src/service/resources-impl.js]: Nit: `element` is guaranteed.

Comment (glevitzky) [reply src/service/resources-impl.js]: Done.

File: src/service/resources-impl.js [hunk 4]
```diff
@@ -1585,6 +1595,27 @@ export class Resources {
...
+   */
+  isWidthOnlyReflowFreeExpansion_(element, width) {
+    const parent = element.parentElement;
+    // If the element has siblings, it's possible that a width-expansion will
+    // cause some of them to be pushed down.
+    if (!parent || parent.childElementCount > 1) {
+      return false;
+    }
+    const parentWidth =
+      (parent.getImpl && parent.getImpl().getLayoutWidth()) || -1;
```
Comment (choumx) [inline src/service/resources-impl.js]: `getImpl()` is async. #24249

Comment (glevitzky) [reply src/service/resources-impl.js]: Not sure how I missed this. Will fix asap!

**Pull request merged**